pkg/storegateway/indexheader/encoding/factory.go

```diff
@@ -154,3 +160,58 @@ func (df *DecbufFactory) CloseWithErrCapture(err *error, d Decbuf, format string
 
 	*err = merr.Err()
 }
+
+// filePool maintains a pool of file handles up to a maximum number, creating
+// new handles and closing them when required. get and put operations on this
+// pool never block. If there are no available file handles, one will be created
+// on get. If the pool is full, the file handle is closed on put.
+type filePool struct {
+	path    string
+	handles chan *os.File
+}
+
+// newFilePool creates a new file pool for path with cap capacity. If cap is 0,
+// get always opens new file handles and put always closes them immediately.
+func newFilePool(path string, cap uint) *filePool {
+	return &filePool{
+		path: path,
+		// We don't care if cap is 0 which means the channel will be unbuffered. Because
+		// we have default cases for reads and writes to the channel, we will always open
+		// new files and close file handles immediately if the channel is unbuffered.
+		handles: make(chan *os.File, cap),
+	}
+}
+
+// get returns a pooled file handle if available or opens a new one if there
+// are no pooled handles available.
+func (p *filePool) get() (*os.File, error) {
+	select {
+	case f := <-p.handles:
+		return f, nil
+	default:
+		return os.Open(p.path)
+	}
+}
+
+// put returns a file handle to the pool if there is space available or closes
+// the file handle if there is not.
+func (p *filePool) put(f *os.File) error {
+	select {
+	case p.handles <- f:
+		return nil
+	default:
+		return f.Close()
+	}
+}
+
+// stop closes all pooled file handles.
+func (p *filePool) stop() {
+	for {
+		select {
+		case f := <-p.handles:
+			_ = f.Close()
+		default:
+			return
+		}
+	}
+}
```

Comment:
I think we need to set a "stopped" flag to make sure that `put` doesn't put more handlers after we've stopped.

```suggestion
func (p *filePool) put(f *os.File) error {
        if p.stopped.Load() {
                return f.Close()
        }
	select {
	case p.handles <- f:
		return nil
	default:
		return f.Close()
	}
}

// stop closes all pooled file handles.
func (p *filePool) stop() {
        p.stopped.Set(true)
	for {
		select {
		case f := <-p.handles:
			_ = f.Close()
		default:
			return
		}
	}
}
```

With `p.stopped` being an `atomic.Bool` or something like that.

Reply:
That makes sense to me but I was holding off doing that because the callers of this code will only call `Reader.Close()` on an index-header reader that has no inflight requests. However, it probably makes sense to ensure this code can't be used incorrectly.

Reply:
Unfortunately, after reading it twice, my code doesn't solve the race condition.

Reply:
Yeah, I can't think of a good way to protect against file handles being added to `handles` after the pool is already "stopped" besides using a `sync.Mutex` in addition to a `stopped` bool.

Reply:
WDYT about this?

https://github.com/grafana/mimir/commit/ce0c44707ff73cd26dda82fb9a9c368685a96292

Reply:
I implemented the solution with a `RWMutex`. I think it will end up being a bit faster than using a `Mutex` because most of the calls to the pool (`get` and `put`) will only need to hold the read lock. I reran the benchmark from above and the results seem about the same:

```
$ benchstat -delta-test=none old.txt new.txt 
name                                   old time/op    new time/op    delta
DecbufFactory_NewDecbufAtUnchecked-16    2.98µs ± 0%    0.91µs ± 0%  -69.49%

name                                   old alloc/op   new alloc/op   delta
DecbufFactory_NewDecbufAtUnchecked-16      232B ± 0%       48B ± 0%  -79.31%

name                                   old allocs/op  new allocs/op  delta
DecbufFactory_NewDecbufAtUnchecked-16      4.00 ± 0%      1.00 ± 0%  -75.00%
```

Reply:
You're right, most of the calls will be for the reads. This way we can use the defer (otherwise we'd be waiting until files are opened or closed on an exclusive mutex).

I think we could do it without the bool by just replacing the chan by a nil one, but it's a matter of tastes.

LGTM.